Field sensitive goto-symex

tautschnig opened the pull request:
~~This is the rebased version of an old patch collection. It's likely broken at present, but should be a starting point for discussing and evaluating this.~~

@smowton Has put a lot of work into this, and this PR now reflects our combined effort. It is now ready for review.

martin-cs commented:
+1 #doeswant

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+    Forall_operands(it, expr)
+      operator()(ns, *it, write);
+
+  if(expr.id()==ID_symbol &&
```
`is_ssa_expr`

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+  bool write) const
+{
+#if 1
+  if(expr.id()!=ID_address_of)
```
Comment here why address-of is special?

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+  {
+    member_exprt &member=to_member_expr(expr);
+
+    if(member.struct_op().id()==ID_symbol &&
```
is_ssa_expr

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+
+      member.struct_op()=tmp.get_original_expr();
+
+      tmp.type()=member.type();
```
Does `tmp` retain any aspect of the original struct-op here?

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+      tmp.set(ID_expression, member);
+      tmp.update_identifier();
+
+      expr.swap(tmp);
```
Definitely need a comment here about what this gymnastic manoeuvre achieves in sum! I think... `some_struct_symbol#n.somefield` -> `some_struct_symbol.member#m`?

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+  }
+  else if(write && expr.id()==ID_index)
+  {
+    index_exprt &index=to_index_expr(expr);
```
`index` -> `index_expr`, since we also have the actual index in play here

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+#endif
+}
+
+/*******************************************************************\
```
Remove old-style comment blocks

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+       to_array_type(ns.follow(index.array().type())).size().id()==ID_constant &&
+       !to_array_type(ns.follow(index.array().type())).size().is_zero())
+    {
+      ssa_exprt tmp=to_ssa_expr(index.array());
```
Comment somewhere about how constant and non-constant array-index operations interact

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+      dest.copy_to_operands(tmp_dest);
+    }
+  }
+  else if(followed_type.id()==ID_array &&
```
Since you use this in a few places, is-constant-sized-array seems like a useful utility predicate

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+
+\*******************************************************************/
+
+void field_sensitivityt::operator()(
```
Since this is being called by name as "operator()" quite a lot, suggest giving it a real name

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+        ++it, ++number)
+    {
+      member_exprt member_rhs(lhs, it->get_name(), it->type());
+      exprt member_lhs=lhs_fs.operands()[number];
```
const ref

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+    for(unsigned i=0; i<array_size_u; ++i)
+    {
+      index_rhs.index()=from_integer(i, index_type());
+      exprt index_lhs=lhs_fs.operands()[i];
```
const ref

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,346 @@
+      field_assignments_rec(ns, state, target, member_lhs, member_rhs);
+    }
+  }
+  else if(followed_type.id()==ID_array)
```
Worth retaining the ability to represent a whole-array assignment compactly? String constants are the main thing that springs to mind -- then break it down into smaller pieces once the program starts making individual array cell assignments?

smowton commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,52 @@
+
+/*! \brief Control granularity of object accesses
+*/
+class field_sensitivityt
```
Since this class maintains no state, these could just be free functions

smowton commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -227,11 +227,117 @@ void goto_symext::symex_assign_symbol(
     tmp_ssa_rhs.swap(ssa_rhs);
   }
 
+  state.field_sensitivity(ns, ssa_rhs, false);
```
Unclear whether this does anything or just accrues state -- give the function name a verb

smowton reviewed (commented):
Some initial comments

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,254 @@
+
+ Purpose:
+
+\*******************************************************************/
```
doxygen (entire PR)

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,254 @@
+    dest=struct_exprt(ssa_expr.type());
+    dest.reserve_operands(components.size());
+
+    for(struct_union_typet::componentst::const_iterator
```
ranged for (entire PR)

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,254 @@
+    const struct_union_typet::componentst &components=
+      type.components();
+
+    assert(components.empty() ||
```
INVARIANT (entire PR)

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,254 @@
+           (lhs_fs.has_operands() &&
+            lhs_fs.operands().size()==components.size()));
+
+    unsigned number=0;
```
size_t

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,254 @@
+    for(exprt::operandst::const_iterator it=lhs.operands().begin();
+        it!=lhs.operands().end();
+        ++it, ++fs_it)
+      field_assignments_rec(ns, state, target, *fs_it, *it);
```
braces

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,52 @@
+
+\*******************************************************************/
+
+#ifndef CPROVER_FIELD_SENSITIVITY_H
```
wrong header guard

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,52 @@
+class goto_symex_statet;
+class symex_targett;
+
+/*! \brief Control granularity of object accesses
```
///

peterschrammel commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -52,14 +52,37 @@ void goto_symext::symex_dead(statet &state)
+    // L2 renaming
+    if(state.level2.current_names.find(l1_identifier)!=
+       state.level2.current_names.end())
+      state.level2.increase_counter(l1_identifier);
```
braces

peterschrammel commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -52,14 +52,37 @@ void goto_symext::symex_dead(statet &state)
+    assert(state.dirty);
+    if(state.dirty->is_dirty(ssa_lhs.get_object_name()) &&
+       state.atomic_section_id==0)
+      target.shared_write(
```
braces

peterschrammel commented on src/goto-symex/symex_main.cpp:
```diff
@@ -404,6 +404,10 @@ void goto_symext::symex_step(
       symex_function_call(get_goto_function, state, deref_code);
+
+      if(deref_code.lhs().is_not_nil())
+        state.field_sensitivity.field_assignments(
```
braces

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -8,6 +8,9 @@ Author: Michael Tautschnig, mt@eecs.qmul.ac.uk
 #include <util/std_expr.h>
+#include <util/arith_tools.h>
+
+#include <ansi-c/c_types.h>
```
goto-symex must not depend on ansi-c

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -120,6 +153,40 @@ void field_sensitivityt::get_fields(
+    const array_typet &type=to_array_type(followed_type);
+
+    mp_integer array_size;
+    if(to_integer(type.size(), array_size))
```
use `numeric_cast`

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -120,6 +153,40 @@ void field_sensitivityt::get_fields(
+    mp_integer array_size;
+    if(to_integer(type.size(), array_size))
+      assert(false);
+    unsigned array_size_u=integer2unsigned(array_size);
```
size_t?

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,254 @@
+  exprt &expr,
+  bool write) const
+{
+#if 1
```
Shall we have command line options to turn on/off field-sensitivity for structs/arrays to make it easier to perform experiments with it?

peterschrammel commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -227,11 +227,73 @@ void goto_symext::symex_assign_symbol(
+  }
+  #endif
+
+  std::cerr << from_expr(ns, "", ssa_rhs) << std::endl;
```
deactivate debug output

peterschrammel commented on regression/cbmc/Array_Propagation1/test.desc:
```diff
@@ -0,0 +1,9 @@
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+\(Starting CEGAR Loop\|VCC(s), 0 remaining after simplification$\)
```
What is the `CEGAR Loop` here? ^^

peterschrammel commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -234,6 +234,50 @@ void goto_symext::symex_assign_symbol(
 
   ssa_exprt lhs_mod=lhs;
 
+  if(ssa_rhs.id()==ID_byte_update_little_endian ||
```
Is there a test that triggers this case?

owen-mc-diffblue commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -239,7 +240,105 @@ void goto_symext::symex_assign_symbol(
+
   do_simplify(ssa_rhs);
 
   ssa_exprt ssa_lhs=lhs;
```
```suggestion
  ssa_exprt ssa_lhs = lhs_mod;
```
This makes regression/cbmc/Anonymous_Struct2 pass, and a few others.

smowton commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+  /// \param ns: a namespace to resolve type symbols/tag types
+  /// \param state: symbolic execution state
+  /// \param target: symbolic execution equation store
+  /// \param lhs_f: expanded symbol
```
lhs_fs

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -453,6 +458,10 @@ bool goto_symex_statet::l2_thread_read_encoding(
     (!ns.lookup(obj_identifier).is_shared() && !(dirty)(obj_identifier)))
     return false;
 
+  // is it an indivisible object being accessed?
```
Comment here on why in that case the rest of this function is inapplicable

romainbrenguier commented on regression/cbmc/variable-access-to-constant-array/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+THOROUGH
```
Why is this changed?
Separating the changes on tests from the changes on code, and commenting on the reason the test is changed in the commit messages would make it easier to review.

tautschnig replied:
I've added a comment:
```
Field-sensitive encoding of array accesses for an array of 2^16 elements is very
expensive. We might consider some bounds up to which we actually do field
expansion.
```

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+
+void field_sensitivityt::apply(const namespacet &ns, exprt &expr, bool write)
+{
+#ifdef ENABLE_FIELD_SENSITIVITY
```
Can you remove these flags?

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+
+#define ENABLE_FIELD_SENSITIVITY
+
+static bool do_apply = true;
```
Can you remove this static value?

tautschnig replied:
That proved rather painful, but it certainly makes for a much clearer design. A `field_sensitivity` object is now being passed around quite a bit, but that avoids hidden state.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+  /// existing member or index expressions into symbols are performed.
+  /// \param ns: a namespace to resolve type symbols/tag types
+  /// \param expr: an expression to be (recursively) transformed - this
+  /// parameter is both input and output.
```
indent this line

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+      member.struct_op() = tmp.get_original_expr();
+      tmp.set_expression(member);
+
+      expr.swap(tmp);
```
It's difficult to figure out what this piece of code is doing.

tautschnig replied:
I have added comments.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+    const struct_typet &type = to_struct_type(followed_type);
+
+    const struct_union_typet::componentst &components = type.components();
+
```
Skipping one line between each line doesn't help readability.

tautschnig replied:
I have removed them and re-arranged some lines.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+  bool do_apply_bak = do_apply;
+  do_apply = false;
+  field_assignments_rec(ns, state, target, lhs_fs, lhs);
+  do_apply = do_apply_bak;
```
Get rid of this `do_apply` stuff (pass it as an argument if needed).

tautschnig replied:
The argument doesn't work, because we need to communicate this across calls to functions of `goto_symex_statet`. Instead, I have made the class stateful, and make sure that the full class is being passed around.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+      state.source,
+      symex_targett::assignment_typet::STATE);
+  }
+  else if(followed_type.id() == ID_struct)
```
could use `try_dynamic_cast<struct_type>` :cake:

tautschnig replied:
I actually got rid of most of the `ns.follow` uses now.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+      ++number;
+    }
+  }
+  else if(followed_type.id() == ID_array)
```
see :cake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+      lhs.has_operands() && lhs_fs.operands().size() == lhs.operands().size());
+
+    exprt::operandst::const_iterator fs_it = lhs_fs.operands().begin();
+    for(const exprt &op : lhs.operands())
```
:pick: maybe better to use an index rather than updating two iterators

tautschnig replied:
Hmm, no, I'd rather do more of these: repeated `.operands()` access is slow.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+  /// \param target: symbolic execution equation store
+  /// \param lhs_f: expanded symbol
+  /// \param lhs: non-expanded symbol
+  static void field_assignments_rec(
```
Since there is no private field, this function doesn't need to appear in the header file.

tautschnig replied:
Was indeed true, but now that's no longer the case as the class has become stateful.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
   }
 
+  // TODO: done via rename
+  // field_sensitivityt::apply(ns, ssa_rhs, false);
```
The TODO comment is not clear (what has to be done?). Ideally it should be resolved before the PR is merged.

tautschnig replied:
Remove, actually the call to `rename` is much clearer now as it takes an additional `field_sensitivity` parameter.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+    lhs_mod = to_ssa_expr(fs_lhs);
+  }
+#endif
+
```
this function is getting too long

tautschnig replied:
Refactored.

romainbrenguier commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -386,11 +387,18 @@ void goto_symext::dereference(
+  rename_l1_reads_l2(expr, state, ns, write);
+  // really, we need to use another dereferencing implementation,
+  // such as dereferencet, with L2 renaming on demand
+#endif
```
remove the commented code

tautschnig replied:
Done.

romainbrenguier reviewed (commented):
I've only partially reviewed the first commit and I think it should be split into several commits.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -17,6 +17,7 @@ Author: Michael Tautschnig
 #include "symex_target.h"
 
 #define ENABLE_FIELD_SENSITIVITY
+// #define ENABLE_ARRAY_FIELD_SENSITIVITY
```
The commit message doesn't explain *why* this change is made

peterschrammel replied:
Is there an issue that tracks resolving this problem?

romainbrenguier commented on src/util/ssa_expr.cpp:
```diff
@@ -31,8 +31,8 @@ static void build_ssa_identifier_rec(
-    os << '.' << member.get_component_name();
-    l1_object_os << '.' << member.get_component_name();
+    os << ".." << member.get_component_name();
+    l1_object_os << ".." << member.get_component_name();
```
This should be documented somewhere. In particular is the documentation of `build_ssa_identifier_rec` still correct?

tautschnig replied:
Documentation updated, thanks for spotting this!

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -31,14 +31,16 @@ class field_sensitivityt
   /// \param write: set to true if the expression is to be used as an lvalue.
-  static void apply(const namespacet &ns, exprt &expr, bool write);
+  static void apply(
+    const namespacet &ns, exprt &expr, bool write, goto_symex_statet &state);
```
The documentation should be updated for the new parameter :duck:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -31,14 +31,16 @@ class field_sensitivityt
   /// \return Expanded expression
-  static exprt get_fields(const namespacet &ns, const ssa_exprt &ssa_expr);
+  static exprt get_fields(
+    const namespacet &ns, const ssa_exprt &ssa_expr, goto_symex_statet &state);
```
:duck:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -60,7 +62,8 @@ class field_sensitivityt
   /// SSA expression.
-  static bool is_indivisible(const namespacet &ns, const ssa_exprt &expr);
+  static bool is_indivisible(
+    const namespacet &ns, const ssa_exprt &expr, goto_symex_statet &state);
```
:duck:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -80,7 +81,8 @@ class field_sensitivityt
     const exprt &lhs_fs,
-    const exprt &lhs);
+    const exprt &lhs,
+    bool allow_pointer_unsoundness);
```
:duck:

tautschnig replied:
Done.

romainbrenguier reviewed (commented):
Some high-level comments. I still need to get a deeper understanding of the changes.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+      {
+        index_exprt &idx = to_index_expr(be);
+
+#ifdef USE_UPDATE
```
If you know what the USE_UPDATE flag means could you document it at the beginning of the file?

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+        new_rhs.new_value() = ssa_rhs;
+#else
+        exprt new_rhs(ID_with, idx.array().type());
+        new_rhs.copy_to_operands(idx.array(), idx.index(), ssa_rhs);
```
use `with_exprt(idx.array(), idx.index(), ssa_rhs)`

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+        new_rhs.copy_to_operands(idx.array(), idx.index(), ssa_rhs);
+#endif
+
+        ssa_rhs = new_rhs;
```
`new_rhs` isn't used afterwards, so assign to `ssa_rhs` directly when it is constructed

peterschrammel replied:
Doesn't seem possible because `ssa_rhs` is used in constructing `new_rhs`.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+    ssa_rhs.id() == ID_byte_update_little_endian ||
+    ssa_rhs.id() == ID_byte_update_big_endian)
+  {
+    byte_update_exprt &bu = to_byte_update_expr(ssa_rhs);
```
avoid abbreviated names, `byte_update` would be better :snake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+    ssa_rhs.id() == ID_byte_update_big_endian)
+  {
+    byte_update_exprt &bu = to_byte_update_expr(ssa_rhs);
+    exprt be = byte_extract_exprt(
```
:snake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+  while(ssa_rhs.id() == ID_update &&
+        (lhs_mod.type().id() == ID_array || lhs_mod.type().id() == ID_struct))
+  {
+    exprt fs_lhs;
```
:snake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+    exprt fs_lhs;
+    const update_exprt &u = to_update_expr(ssa_rhs);
+    assert(u.designator().size() == 1);
+    const exprt &d = u.designator().front();
```
:snake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+        (lhs_mod.type().id() == ID_array || lhs_mod.type().id() == ID_struct))
+  {
+    exprt fs_lhs;
+    const update_exprt &u = to_update_expr(ssa_rhs);
```
:snake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+  while(ssa_rhs.id() == ID_with &&
+        (lhs_mod.type().id() == ID_array || lhs_mod.type().id() == ID_struct))
+  {
+    exprt fs_lhs;
```
:snake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -231,10 +232,108 @@ void goto_symext::symex_assign_symbol(
+        (lhs_mod.type().id() == ID_array || lhs_mod.type().id() == ID_struct))
+  {
+    exprt fs_lhs;
+    const with_exprt &w = to_with_expr(ssa_rhs);
```
:snake:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+class symex_targett;
+
+/// \brief Control granularity of object accesses
+class field_sensitivityt
```
This class has no field. Make the functions free.

tautschnig replied:
See above: that's no longer possible.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,251 @@
+      index.array() = tmp.get_original_expr();
+      tmp.set_expression(index);
+
+      expr.swap(tmp);
```
here also it's difficult to see the global effect of this sequence of instructions

tautschnig replied:
Comment added.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+  /// field-sensitive SSA representation of \p ssa_expr.
+  /// \param ns: a namespace to resolve type symbols/tag types
+  /// \param expr: the expression to evaluate
+  /// \return Expanded expression
```
It's not clear from the documentation what the returned expression is.

tautschnig replied:
I have added an example.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+  static exprt get_fields(const namespacet &ns, const ssa_exprt &ssa_expr);
+
+  /// Assign to the individual fields of a non-expanded symbol \p lhs. This is
+  /// required whenever prior steps have updated the full object rather than
```
What is the requirement?

tautschnig replied:
Hmm, I don't really understand what you mean by "requirement" here?

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+    const exprt &lhs);
+
+  /// Determine whether \p expr would translate to a single field-sensitive SSA
+  /// expression.
```
What do you mean by single field-sensitive SSA? How can it translate to several?

tautschnig replied:
Documentation expanded.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+  /// \param state: symbolic execution state
+  /// \param target: symbolic execution equation store
+  /// \param lhs_f: expanded symbol
+  /// \param lhs: non-expanded symbol
```
Since this is not part of the interface, the documentation should go to the cpp file.

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,83 @@
+  /// \param expr: the expression to evaluate
+  /// \return True, if and only if, \p expr would be a single field-sensitive
+  /// SSA expression.
+  static bool is_indivisible(const namespacet &ns, const ssa_exprt &expr);
```
It seems the result is always negated. Can we change that to a `is_divisible` function, to avoid double negations?

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -45,14 +46,34 @@ void goto_symext::symex_dead(statet &state)
+    // prevent propagation
+    state.propagation.erase(l1_identifier);
+
+    // TODO: not sure why the rest below is necessary
```
Then let's remove it. If no test fails then it is not needed, no? ;)

tautschnig replied:
#3980 actually clarifies those comments. (This PR just changes the indent, no actual code changes here.)

peterschrammel commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -377,6 +377,24 @@ void goto_symext::symex_assign_symbol(
+
+    // The constant propagator is indexed by L1 names:
+    state.rename(underlying_composite, ns, goto_symex_statet::L1);
+    state.propagation.erase(underlying_composite.get_l1_object_identifier());
```
Could we update the composite object instead of invalidating it?

peterschrammel commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -63,17 +63,5 @@ void goto_symext::symex_dead(statet &state)
     if(level2_it != state.level2.current_names.end())
       symex_renaming_levelt::increase_counter(level2_it);
-
-    const bool record_events = state.record_events;
```
Putting this reasoning into the code would be good.

tautschnig replied:
What are you referring to by "this reasoning?"

tautschnig commented:
This should be ready for re-review and/or further comments.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -119,6 +125,7 @@ exprt field_sensitivityt::get_fields(
 
     return result;
   }
+#ifdef ENABLE_ARRAY_FIELD_SENSITIVITY
```
Can we use template arguments instead of precompilation flags? It makes the code clearer and easier to test.

tautschnig replied:
I briefly tried, but this is really painful to thread through. So I'll refrain from picking this one up.

romainbrenguier reviewed (commented):
The commits appear in the wrong order on github and it looks like some of them could be squashed.

smowton committed:
Reinstate symex-deref's `write` parameter

field_sensitivity::apply needs to avoid expanding struct-typed symbols
into fields when they're being used as lvalues.

smowton committed:
Field sensitivity: avoid inheriting L2 renaming from struct symbols

Previously this could turn e.g. struct@x#y into struct.field@x#y,
regardless of whether the L2 generation #y for that field existed or
not. This meant that when an L2 name was passed through
field-sensitivity expansion it could end up referring to an unbound
symbol, whereas an L1 name would remain coherent (e.g.  struct@x ->
struct.field@x), later being renamed using the ordinary level2_names
map. That means the generation numbers for structs and their fields may
get out of sync, but that should be harmless so long as "latest
generation" is used consistently.

We can now also remove constant propagator lookup after applying field
sensitivity, because field_sensitivity::apply uses state.rename to apply
L2 renaming, which itself consults the constant propagator.

Also make is_divisible test cheaper, at the expense that synchronisation
with get_fields needs to be maintained manually.

smowton committed:
Temporarily disable field-sensitivity for arrays

The Array_Propagation1 test is no longer solved during symex in this
case.

smowton committed:
Use `..` and `[[]]` for SSA expression field separator

This makes it easier to read the difference between an atomic symbol and
a field of a composite.

smowton committed:
Apply field-sensitivity expansion after removing derefs

This is because a pointer into the middle of a struct can resolve into
something like &struct.field, in which case we need to rewrite that as
the field-sensitive SSA symbol before any further dereferencing.

smowton committed:
Comment on and clarify field-sensitivity-specific expression transformations

The function name was previously incorrect (it did more than it said),
so now I just call both transformations from symex_assign.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8100805058b0228fd0eea45dd9f24e118d8e3a83).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103944749

romainbrenguier commented:
I tried the current version of the branch on some benchmarks and there were a couple of regressions but on the instances which worked the performances where 32% faster than develop (using the BDD compilation options in both case (using BDD compilation options is 30% faster on develop)). So this is promising. I will try to come up with minimal examples of the regressions.

tautschnig commented:
Thank you @romainbrenguier - looking forward to debugging guidance. On SV-COMP I haven't yet observed regressions that I can clearly attribute to field-sensitivity (but certainly also some performance improvement).

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,253 @@
+    return;
+
+  if(expr.id() != ID_address_of)
+    Forall_operands(it, expr)
```
Add braces

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,80 @@
+  /// expressions. Such expansions are not possible when the symbol is to be
+  /// written to (i.e., \p write is true); in such a case only translations from
+  /// existing member or index expressions into symbols are performed.
+  /// \param expr: an expression to be (recursively) transformed - this
```
~\param [in,out]~

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,80 @@
+  /// required whenever prior steps have updated the full object rather than
+  /// individual fields, e.g., in case of assignments to an array at an unknown
+  /// index.
+  /// \param state: symbolic execution state
```
\param [in,out]

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -198,6 +202,141 @@ void goto_symext::symex_assign_rec(
+    PRECONDITION(update.designator().size() == 1);
+    const exprt &designator = update.designator().front();
+
+    if(lhs_mod.type().id() == ID_array)
```
Add braces

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -40,29 +40,30 @@ class field_sensitivityt
   /// Compute an expression representing the individual components of a
   /// field-sensitive SSA representation of \p ssa_expr.
-  /// \param ns: a namespace to resolve type symbols/tag types
+  /// \param state: symbolic execution state
```
\param [in,out]

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -40,29 +40,30 @@ class field_sensitivityt
   /// expressions. Such expansions are not possible when the symbol is to be
   /// written to (i.e., \p write is true); in such a case only translations from
   /// existing member or index expressions into symbols are performed.
+  /// \param state: symbolic execution state
```
\param [in,out]

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -40,29 +40,30 @@ class field_sensitivityt
   /// components (such as a struct with each member becoming an individual SSA
   /// expression, return true in this case).
-  /// \param ns: a namespace to resolve type symbols/tag types
+  /// \param state: symbolic execution state
```
\param [in,out]

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -34,7 +34,18 @@ class field_sensitivityt
+  /// If `is_divisible(lhs)` then assign indivisible symbols (fields or array
+  /// indices) from the divisible symbol and erase the divisible symbol from
+  /// \p state.
+  /// \param state: symbolic execution state
```
\param [in,out]

tautschnig replied:
Done.

peterschrammel commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -181,6 +181,28 @@ void field_sensitivityt::field_assignments(
+    // the two becoming out of sync).
+    // Note we can't use code_deadt, which would actually try to kill the
+    // components, which we *do* want to keep!
+    state.propagation.erase(lhs.get_l1_object_identifier());
```
Could we update the composite object instead of invalidating it?

smowton replied:
I don't think we should, the whole point is this isn't a canonical symbol and we shouldn't be carrying stale versions of it around

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,253 @@
+void field_sensitivityt::apply(exprt &expr, bool write) const
+{
+  if(!run_apply)
+    return;
```
I think the caller should do the check to know whether to call apply

smowton replied:
I think this may no longer be necessary, as `field_assignments_rec` does not use `apply`

tautschnig replied:
`field_assignments_rec` calls `rename`, which turn will call `apply`. And I don't think it should be upon the caller to know whether to call `apply`, because that would mean the implementation details of field sensitivity are spilled all over the place. So I'd really keep this as is.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,253 @@
+    !write && expr.id() == ID_member &&
+    to_member_expr(expr).struct_op().id() == ID_struct)
+  {
+    simplify(expr, ns);
```
Why are we simplifying here?

tautschnig replied:
We have a member expression over a struct expression, so this can be collapsed to just access that member directly. We might ask the question whether this can/does appear at all (as the simplifier should have taken care of this before).

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,80 @@
+  /// \return Expanded expression; for example, for a \p ssa_expr of some struct
+  ///   type, a `struct_exprt` with each component now being an SSA expression
+  ///   is built.
+  static exprt get_fields(const namespacet &ns, const ssa_exprt &ssa_expr);
```
When I see a static member function, I assume it will access some private fields of a `field_sensitivityt` object. It seems it's not the case here, so for me this is misleading.

tautschnig replied:
If we were using namespaces then that would be the right way to go - this is really just about logical coupling. So I'd rather keep it as-is.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -0,0 +1,80 @@
+  /// \param expr: the expression to evaluate
+  /// \return False, if and only if, \p expr would be a single field-sensitive
+  /// SSA expression.
+  static bool is_divisible(const namespacet &ns, const ssa_exprt &expr);
```
same remark as for `get_fields`

tautschnig replied:
As above.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -359,16 +376,23 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
       "if_exprt");
+
+    if(level == L2)
+      field_sensitivity.apply(expr, false);
```
Is it true that field_sensitivity is only used at level 2? In that case it could make sense to have it as a field of `symex_level2t`

tautschnig replied:
I'm looking into this, it might be a great way of reducing the size of this PR. But it'll take me a bit to implement it.

tautschnig replied:
Done - it has indeed reduced the churn a lot.

romainbrenguier commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -31,8 +31,11 @@ Author: Daniel Kroening, kroening@kroening.com
+
 /// \brief Central data structure: state.
 ///
+
```
unrelated change

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -198,6 +202,141 @@ void goto_symext::symex_assign_rec(
       "assignment to `" + lhs.id_string() + "' not handled");
 }
 
+static void rewrite_with_to_field_symbols(
```
This should be documented, in particular the fact there are to arguments which are both output and input seems strange to me.

tautschnig replied:
Documentation added.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -198,6 +202,141 @@ void goto_symext::symex_assign_rec(
+#endif
+}
+
+static void shift_indexed_access_to_lhs(
```
same comment as for `rewrite_with_to_field_symbols`

tautschnig replied:
Documentation added.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -198,6 +202,141 @@ void goto_symext::symex_assign_rec(
+          index_exprt &idx = to_index_expr(byte_extract);
+
+#ifdef USE_UPDATE
+          update_exprt new_rhs(idx.array().type());
```
use braces

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -198,6 +202,141 @@ void goto_symext::symex_assign_rec(
+          new_rhs.designator().push_back(index_designatort(idx.index()));
+          new_rhs.new_value() = ssa_rhs;
+#else
+          with_exprt new_rhs(idx.array(), idx.index(), ssa_rhs);
```
use braces

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -198,6 +202,141 @@ void goto_symext::symex_assign_rec(
+          const irep_idt &component_name = member.get_component_name();
+
+#ifdef USE_UPDATE
+          update_exprt new_rhs(member.compound().type());
```
{}

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -198,6 +202,141 @@ void goto_symext::symex_assign_rec(
+          new_rhs.designator().push_back(member_designatort(component_name));
+          new_rhs.new_value() = ssa_rhs;
+#else
+          with_exprt new_rhs(member.compound(), exprt(ID_member_name), ssa_rhs);
```
{}

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -32,20 +34,28 @@ void goto_symext::symex_dead(statet &state)
-  if(level2_it != state.level2.current_names.end())
-    symex_renaming_levelt::increase_counter(level2_it);
+  const exprt l2_fields = field_sensitivityt::get_fields(ns, ssa);
+  std::set<exprt> l2_fields_set;
```
does it need to be ordered?

tautschnig replied:
Actually the unordered version of `find_symbols` is indeed much nicer here. Cleanup done.

romainbrenguier commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -32,20 +34,28 @@ void goto_symext::symex_dead(statet &state)
+
+  for(const exprt &l2_field_set : l2_fields_set)
+  {
+    ssa_exprt ssa_lhs = to_ssa_expr(l2_field_set);
```
it looks it's only used to get the identifier, so this can be inlined in the `l1_identifier` declaration

tautschnig replied:
Now gone.

romainbrenguier commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -224,7 +225,8 @@ void goto_symext::dereference_rec(exprt &expr, statet &state, bool write)
     // we need to set up some elaborate call-backs
-    symex_dereference_statet symex_dereference_state(state, ns);
+    symex_dereference_statet symex_dereference_state(
+      state, ns, field_sensitivity);
```
{ }

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -68,10 +71,15 @@ void field_sensitivityt::apply(exprt &expr, bool write) const
       // place the entire member expression, not just the struct operand, in an
       // SSA expression
       ssa_exprt tmp = to_ssa_expr(member.struct_op());
```
I think this could be a reference instead of a copy since expr doesn't seem to be used afterwards

tautschnig replied:
I don't think that would be a big improvement.

romainbrenguier commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -414,6 +414,25 @@ void goto_symext::symex_assign_symbol(
+  if(ssa_lhs.get_original_expr().id() == ID_member)
+  {
+    ssa_exprt underlying_composite = ssa_lhs;
+    exprt underlying_object = underlying_composite.get_original_expr();
```
Can you use a reference wrapper instead, to avoid copies in the loop?

tautschnig replied:
This code is gone.

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -270,9 +270,19 @@ void field_sensitivityt::field_assignments_rec(
 {
-  return expr != get_fields(state, expr);
+  if(expr.type().id() == ID_struct || expr.type().id() == ID_struct_tag)
+    return true;
```
This change should be squashed with the first "field sensitivity" commit

tautschnig replied:
Possible, but I'd rather maintain proper attribution (@smowton has contributed these improvements).

romainbrenguier commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -181,6 +181,28 @@ void field_sensitivityt::field_assignments(
+    // components, which we *do* want to keep!
+    state.propagation.erase(lhs.get_l1_object_identifier());
+
+    exprt unknown_expr(ID_unknown, lhs.type());
```
{ } and we should have an `unknown_exprt` type

tautschnig replied:
This is now gone.

romainbrenguier commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -34,7 +34,18 @@ class field_sensitivityt
+  /// \param state: symbolic execution state
+  /// \param lhs: non-expanded symbol
+  /// \param ns: symex's working namespace
+  void post_process_assignment(
```
not a great name, I'm not sure what it should be, something like  `assign_indivisible_and_erase_divisible`?

tautschnig replied:
Does the user actually need to know what it exactly does? The user should know when to invoke it (after the assignment -> post process), but not necessary what exactly is done in there (which might actually change).

romainbrenguier reviewed (commented):
It looks like several commits should be squashed, as some changes are introduced and then undone.
I am also a bit worried by the coupling between field_sensitivity and `goto_symex_statet`.

smowton commented on regression/cbmc/variable-access-to-constant-array/test.desc:
```diff
@@ -1,10 +1,15 @@
 This test actually passes when using the SMT2 back-end, but takes 68 seconds to
 do so.
+
+Field-sensitive encoding of array accesses for an array of 2^16 elements is very
```
Given this PR doesn't do array-cell sensitivity yet, this might no longer be relevant

tautschnig replied:
Well, it does remain true whether or not we do it right now I'd say.

smowton reviewed (approved):
A few small things to fix, giving my approval as I'd be happy for this to be merged and then incrementally improved

martin-cs reviewed (approved):
There is one file that is in `goto-instrument` so I am apparently a blocking reviewer...  I'm fine with that change and trust that @tautschnig and @smowton are doing a good job on the rest in `symex`.

smowton committed:
Remove composite object from the constant propagator when a field is updated

Otherwise e.g. in "memset2" we had the whole object being initialised
with "{0, 0}" (which is stored in const prop and then expanded to
individual fields), then later any operation that needed the
whole-struct representation (such as the memset-of-one-field in that
test, which expanded to "whole_struct = byte_update(whole_struct,
offset, int, 0)") would get the whole-struct definition stored in
const-prop, disregarding any intermediate updates that were able to
directly reference a single field.

This avoids having two different versions of a symbol (the
field-of-composite and the indivisible field versions) in the most
direct way possible.

tautschnig commented:
@peterschrammel @romainbrenguier This should now be a lot easier to review. I have also applied squashing where I thought it was reasonable to do. Please do take another look.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1a72935ad1c94981ac45bc36f1648be49e9b4440).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106087829
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

smowton commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -0,0 +1,321 @@
+    // Note we can't use code_deadt, which would actually try to kill the
+    // components, which we *do* want to keep!
+    //
+    // There is no need to update value sets, because they are field sensitive
```
I think we should -- as it stands the first assignment will create `a.field1 = ..., a.field2 = ...` and then the second one will move those over to `a..field1` and `a..field2`. That *should* be harmless, as any future read should refer to the `..` field-sensitive versions, but it's at least a waste of memory and at worst going to make broken code (which somehow accidentally refers to the composite versions) appear is if it works

tautschnig replied:
But this just hits https://github.com/diffblue/cbmc/blob/f8dff52944742f7777ad763990eb101ede862b7c/src/pointer-analysis/value_set.cpp#L1203 - we could remove the `UNREACHABLE`, but it wouldn't have any effect either way?!

smowton replied:
I think you'd need to add a method for this -- `erase_symbol` or something which would special-case struct-typed ones to drop each field, similar to how `assign` special-cases struct-typed values.

tautschnig replied:
Done (by adding `state.value_set.values.erase(l1_identifier);`).

smowton replied:
That won't work (the value-set processes assignments to `composite` by decomposing them into assignments to `composite.field1`, `composite.field2`, ....). Suggest making a method on `value_sett` which applies the same decomposition logic is `value_sett::assign`

tautschnig replied:
So I reworked this a bit (getting rid of `post_process_assignment` altogether, but removing from the value set is actually not possible! Doing so breaks several C tests (it doesn't break any Java tests).

tautschnig replied:
Just to clarify what I experimented with: take a backup copy of `value_set` before `goto_symex_statet::assignment` adds information to `value_set` and restore the value set after having done field assignments.

tautschnig replied:
That copy had not taken into account the field assignments happening in between. That's now fixed, all happening in `goto_symex_statet::assignment`.

smowton reviewed (approved):
One last request...

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fda57d000a09c6e8bf2c8df266ba750ef2329821).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106153721
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1708b9057597a60abad4e0bb961ae299f784fbdf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106352811
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5a5be49083996b8503a124bbe87fd67e6463a4ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106386612

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ec20a714bbe9590bfa25d5ce24604d08b9114f11).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106426448

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fbf7d7fa05af230291847885de36f6554a55527f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107048557
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 78f75c773edf7032d4310d4c35660684b9bef3ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107051561

tautschnig committed:
Field-sensitive level-2 SSA renaming

Struct member and array index expressions are replaced by new SSA
symbols such that updates to individual fields of structs or arrays do
not affect the SSA index of the remaining object. This enables more
fine-grained constant propagation, which should be particularly
beneficial to Java (where structs/classes prevail).

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: b38a424c6db25a58733fd3a14e9e1c7aecc267f9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107119069
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 28293c705657e580591254f69146512fa62a2415).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107327902
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

smowton committed:
Move field expansion outside goto_symex_statet::assignment

Otherwise while everything works correctly for the state, the target
sees the field expansions before the composite definition, which is a
problem if the values are non-constant and so the expansion RHSes refer
to the composite!

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7f16d2533ea43c99d9639e977a974fcfcdc4892b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107604152
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 4c236255b10c7cbe12dde828598c8acf4a176c8d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107608202
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 092d8fbbb31ab78536fa4b4c901fedc8d1b6ae4f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107795688

smowton committed:
field_sensitivityt::get_fields: recurse on operands even when input is already L2

For an L1 or lower expression, get_fields already recursed, e.g. to turn
`x` -> `{ { x..a..g, x..a..h }, x..b }` rather than `{ x..a, a..b }`
when its `a` member is itself struct-typed. However this recursion was
omitted in the case where the input expression is already L2 renamed.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5b86abaf2546fb484d1573fa552d3bd18ba67cb9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108113827

romainbrenguier commented:
Here are the results on some benchmarks. Field sensitivity seems to scale much better. Total time was 
76.41 instead of 98.24. However two benchmarks failed because of invariant violations, I will try to produce a minimal example for these failures. 


![chart (1)](https://user-images.githubusercontent.com/11854194/56041994-f3687280-5d31-11e9-9eea-9405c435826a.png)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 098c0cbab6b1ca2c2026b222575344c6945a9a87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108161278

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9329cc2265397c573d954d9c6774eb16558d30e7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108198940

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5dd0b3a07aca874ce548bcc7e15c28080551fab5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108226499

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8cd443cf14eec1da3a062de983665772239d00fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108538361
